Sender Initiated Reference ID Exchange (#186)
diem/dip: dips/dip-183.md

---
dip: 183
title: Sender Initiated Reference ID Exchange
authors: Sunmi Lee (@sunmilee), David Wolinsky (@davidiw)
status: Draft
type: Informational
created: 06/08/2021
last_updated: 06/08/2021
issue: https://github.com/diem/dip/issues/183
---

import useBaseUrl from '@docusaurus/useBaseUrl';

# Summary

This DIP describes a protocol standard for sender initialized pre-flight exchange using reference ID to form payment transactions.

> Reference ID is used as a unique identifier for a payment transaction between two parties across two VASPs without revealing any information about those parties on-chain.

Reviewer: IIUC, each VASP will need to maintain every unique (reference id, sending VASP, receiving VASP) identifier to ensure avoiding duplicate reference ids?

Can VASP A initiate a reference ID R to VASP B
and VASP B also legally initiate a reference ID R to VASP A or is it unique per pair?

Author: Unique per pair, so between VASP A and B, if a reference ID R was used in a previous transaction, regardless of which one initiated it, R cannot be reused for any future transactions between A and B

Reviewer: Thanks for the clarification.  I think it would be good to add something like your statement into the DIP to make this clear.

We should consider from an implementor perspective how feasible this is to implement.  I guess it is expected that the implementor would need to maintain an index of (canonical(VASP1, VASP2), reference ID) and rebuild it from the event history if they lost the index.

This might be simpler if, say, there was just a monotonically increasing reference ID number so one could just look at the last number between canonical(VASP1, VASP, last reference ID number) event on-chain rather than keeping an index?

Author: I think it depends on how the VASP decides to implement it. They can always enforce uniqueness across all reference IDs instead of pair, which is what I do in my miniwallet diem ID example.

created: 06/08/2021
last_updated: 06/08/2021
issue: https://github.com/diem/dip/issues/183
---

import useBaseUrl from '@docusaurus/useBaseUrl';

# Summary

This DIP describes a protocol standard for sender initialized pre-flight exchange using reference IDs to form payment transactions.

A reference ID is used as a unique identifier for a payment transaction between two parties across two VASPs without revealing any information about those parties on-chain.
In this protocol, the sender already knows the recipient VASP address and an identifier for the recipient at their VASP. The sender VASP contacts the recipient VASP and shares the sender's identifier, a reference ID, and the recipient's identifier. 
Upon conclusion, the sender VASP transmits to the blockchain a transaction with the reference ID, completing the protocol without revealing any potentially identifiable user information publicly on-chain.

Reference IDs are unique between a pair of VASPs, `VASP A` and `VASP B`,  so if a reference ID `r` was used in a previous transaction between the two VASPs, regardless of which one initiated it, `r` cannot be reused for any future transactions between `VASP A` and `VASP B`. 

> However, the same `r` value can be used between a different VASP pair, `VASP A` and `VASP C`. VASPs may choose to keep reference IDs globally unique across all VASPs as well. 

Author: Also added more details on ref ID uniqueness @aching

Reviewer 2: Personally I think the first sentence was good enough, but I'd rather over specify than have confusion :P

    },
    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a",
}
```

**ReferenceIDCommandResultObject:**

| Field 	    | Type 	     | Required? 	| Description 	           |
|-------	    |------	     |-----------	|-------------	           |
| _ObjectType   | str        | Y | Fixed value: `SenderInitiatedReferenceIDResponse`|
| receiver_address       | str | Y | Receiver's onchain [account identifier](https://github.com/diem/dip/blob/main/dips/dip-5.md) with subaddress set to `None` or the zero subaddress |

#### Error Codes

`duplicate_reference_id`: Duplicate Reference ID was rejected by the receiving end

`invalid_receiver`: Receiving end could not find the user with the given user_identifier

`invalid_field_value`: Reference ID is in the wrong format. See more details in [DIP-1](https://github.com/diem/dip/blob/main/dips/dip-1.mdx#request-object-validation-error-codes)

## On-Chain Transaction Settlement
If the amount is below the travel rule limit, the sending VASP can send a p2p transaction with PaymentMetadata and the `reference_id` VASPs agreed on to settle the transaction.
```
enum PaymentMetadata {
    PaymentMetadataV0(ReferenceId),
}
type ReferenceId = [u8, 16];
```

> If the amount is at or above the travel rule limit, VASPs should trigger an off-chain protocol for KYC exchange. The same `reference_id` must be used to perform a travel rule exchange as described in [DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.md).

Reviewer 2: ```suggestion
If the amount is at or above the travel rule limit, VASPs must perform an off-chain identity exchange. The same `reference_id` must be used to perform a travel rule exchange as described in [DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.md).
```
